modules/light-clients/08-wasm/keeper/keeper.go

```diff
@@ -29,19 +27,15 @@ type Keeper struct {
 	authority string
 }
 
-// NewKeeper creates a new NewKeeper instance
-func NewKeeper(cdc codec.BinaryCodec, key storetypes.StoreKey, authority string) Keeper {
-	// Wasm VM
-	const wasmDataDir = "ibc_08-wasm_client_data"
-	wasmSupportedFeatures := strings.Join([]string{"storage", "iterator"}, ",")
-	wasmMemoryLimitMb := uint32(math.Pow(2, 12))
-	wasmPrintDebug := true
-	wasmCacheSizeMb := uint32(math.Pow(2, 8))
-
-	vm, err := cosmwasm.NewVM(wasmDataDir, wasmSupportedFeatures, wasmMemoryLimitMb, wasmPrintDebug, wasmCacheSizeMb)
-	if err != nil {
-		panic(err)
-	}
+// NewKeeper creates a new NewKeeper instance with the provided Wasm VM.
+// This constructor function is meant to be used when the chain uses x/wasm
+// and the same Wasm VM instance should be shared with it.
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	key storetypes.StoreKey,
+	authority string,
+	vm *cosmwasm.VM,
+) Keeper {
 	types.WasmVM = vm
```

Comment:
Because this function needs to be used in a very specific way, I think it might be beneficial to deviate from the standard "NewKeeper" naming convention. Maybe `NewKeeperWithSharedVM` to be extra explicit about it.

I think it would be nice to do a check for

```golang
if types.WasmVM != nil {
   panic("meaningful message")
}
```

This at least prevents accidental re-assignment of the global vm when this use case.

Reply:
I could rename it to `NewKeeperWithVM`, since the VM could be or not shared.

Regarding the check, why wouldn't we want to re-assign the global VM? If the global VM is set, that means that either it was set in the keeper constructor function or outside of it, and this should not be standard behaviour, so wouldn't it be better to just re-assign?

Reply:
I'd like `NewKeeperWithVM`, with `Shared` sounds scarier than it should be in my mind (and would also make the call to it in `NewKeeperWithConfig` slightly confusing :smile:

> why wouldn't we want to re-assign the global VM?

I'm unsure of the implications of allowing re-assigning, it seems like it would mess all the internals of the VM up? Definitely sounds like something we should at least look into (i.e ask confio or look at what wasmd does, if anything)

Reply:
My thinking is that presumably if it is creating and set outside of the function, it will be the expectation of the dev that the configure vm will be used by the module. This VM would potentially be ignored if a different one is passed in. While not likely to happen, I think it makes sense to explicitly fail if the VM is configured in a non standard way. This way the only way the VM can be configured is via this function when there is not a VM already configured.

My understanding of the doc string in this PR is that this function should only be used when a shared VM is what is required. I think this check would slightly reduce the surface area for possible issues.

Please correct me if my understanding is incorrect!

Reply:
Panicking mucks up our tests which create a bunch of test chains. I think this should be changed to only panic if the VM that we're trying to set (if set) is different than what we already have. I'll push that change and open an issue to keep track of this and discuss at later point.